### PR TITLE
[docs] adds simple bridge networking documentation.

### PR DESCRIPTION
**Pull request by mccurdyc**
Signed-off-by: Colton J. McCurdy <replace>

# Reason for This PR

https://github.com/firecracker-microvm/firecracker/issues/2618

## Description of Changes

Adds simple bridge networking documentation.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [ ] ~Any newly added `unsafe` code is properly documented.~
- [ ] ~Any API changes are reflected in `firecracker/swagger.yaml`.~
- [ ] ~Any user-facing changes are mentioned in `CHANGELOG.md`.~
- [x] All added/changed functionality is tested.

### CODE REVIEW

**Comment by mccurdyc** [comment]
Force pushed to fix unverified commits due to a broken git config on this new machine.

File: docs/network-setup.md [hunk 1]
```diff
@@ -43,6 +43,46 @@ those rules before starting.
 sudo iptables-save > iptables.rules.old
 ```
 
+*Advanced:*  In case you want to use a bridge to communicate with the microVM,
+you need to do the following additional setup:
```
Comment (serban300) [inline docs/network-setup.md:47]: A part of this steps don't look like additional setup. They overlap with the basic tap setup. For example creating the tap interface, setting it up, etc.

File: docs/network-setup.md [hunk 2]
```diff
@@ -96,6 +136,52 @@ your environment. For testing, you can add a public DNS server to
...
+   ```
+
+1. Setup local DNS via `/etc/hosts`
+   ```
+   # cat /etc/hosts
+   127.0.0.1 localhost
+   ::1 localhost
+   127.0.1.1   <your guest's hostname>.localdomain
+   ```
+
```
Comment (serban300) [inline docs/network-setup.md:184]: This doesn't look specific to the bridge setup.

**Review by serban300 — COMMENTED** [review]
Thanks for the contribution ! Functionally the PR look good, but I find it quite hard to follow the steps because of the way they are mixed as an advanced section inside the current document structure. Also it's very hard to understand what part of the setup is bridge specific, what is tap specific and what is general setup. Could we restructure the document somehow in order to have a clearer separation between the general steps, the tap specific steps and the bridge specific steps ?

**Comment by mccurdyc** [comment]
@serban300 

> Could we restructure the document somehow in order to have a clearer separation between the general steps, the tap specific steps and the bridge specific steps?

Yeah! What were you thinking? A separate "Bridge networking" sub-section (H3) under an "Advanced Section" (H2) at the bottom of the doc? And leaving the tap creation separate, as it is already documented?

I may have misinterpreted [@dianpopa's comment](https://github.com/firecracker-microvm/firecracker/issues/2618#issuecomment-895292914) in the related issue.

**Comment by dianpopa** [comment]
> @serban300
> 
> > Could we restructure the document somehow in order to have a clearer separation between the general steps, the tap specific steps and the bridge specific steps?
> 
> Yeah! What were you thinking? A separate "Bridge networking" sub-section (H3) under an "Advanced Section" (H2) at the bottom of the doc? And leaving the tap creation separate, as it is already documented?
> 
> I may have misinterpreted [@dianpopa's comment](https://github.com/firecracker-microvm/firecracker/issues/2618#issuecomment-895292914) in the related issue.

@mccurdyc you did not misinterpret my comment. Mu hunch on the steps you provided was that the bridge setup makes use of the already existent tap interface setup so I did not see any reason for adding a new section that would replicate the already existent interface setup. @serban300 you say we could have "a clearer separation between the general steps, the tap specific steps and the bridge specific steps". My impression is that the bridge steps also contain the tap steps, am I mistaken?

**Comment by mccurdyc** [comment]
@dianpopa actually, I think I could build on the existing steps for setting up the tap interface.

The only bit that was unclear to me is that the guest's interfaces and routes will need to be setup differently based on whether it is part of a bridge or not.

However, the host tap interface would be the same other than not assigning an IP to the interface and instead, just adding it to the bridge interface.

**Comment by mccurdyc** [comment]
I can try to tweak my docs this evening and we can reassess and determine whether it moves in the correct direction or if it is worse :smile:

**Review by serban300 — COMMENTED** [review]
Thanks for addressing the comments. It looks good to me. Please just squash your commits.

**Comment by mccurdyc** [comment]
@serban300 thanks! Done!

**Comment by serban300** [comment]
@mccurdyc can you also remove the part about addressing my comments from the description please ? I think just the title of the commit + the signed-off-by line is enough.

**Comment by mccurdyc** [comment]
@serban300 yep! Done!

**Review by serban300 — DISMISSED** [review]

**Review by dianpopa — DISMISSED** [review]

**Review by dianpopa — COMMENTED** [review]
The style checks are failing on this PR. Take a look at them and let us know if we can clarify anything.

**Comment by mccurdyc** [comment]
@dianpopa @serban300 I've addressed CI issues. Thanks!

File: docs/network-setup.md [hunk 3]
```diff
@@ -96,6 +96,85 @@ your environment. For testing, you can add a public DNS server to
 nameserver 8.8.8.8
 ```
 
+## [Advanced] Setting Up a Bridge Interface
+
+<!-- markdownlint-disable-next-line MD024 -->
```
Comment (serban300) [inline docs/network-setup.md:101]: What was the warning here ? Can we fix it instead of disabling it ?

Comment (mccurdyc) [reply docs/network-setup.md:101]: I don't see it failing in the [original failed CI build](https://buildkite.com/firecracker/firecracker-x86-64-amd-ci/builds/3166), but locally with markdownlint, the error that I was getting was related to using a duplicate heading because it would create links like `#on-the-host-1`

I can try to fix this this evening.

Comment (mccurdyc) [reply docs/network-setup.md:101]: The reason I chose to disable the lint instead of changing the header was to maintain the same section titling as the normal  --- i.e., non-bridge --- section.

Comment (dianpopa) [reply docs/network-setup.md:101]: We are already disabling this rule [here](https://github.com/firecracker-microvm/firecracker/blob/main/tests/host_tools/mdl_style.rb#L6), so there is no need to do it here.

File: docs/network-setup.md [hunk 4]
```diff
@@ -96,6 +96,85 @@ your environment. For testing, you can add a public DNS server to
...
+1. Add firewall rules to allow traffic to be routed to the guest
+
+   ```bash
+   sudo iptables -t nat -A POSTROUTING -o br0 -j MASQUERADE
+   ```
+
+### On The Guest
+
+1. Define an unused IP address --- in the bridge's subnet e.g., `192.168.1.0/24`
+for the guest's ethernet interface. This is also [documented above](#on-the-host).
```
Comment (dianpopa) [inline docs/network-setup.md:135]: Bring this line to under the one above and the CI should pass.

Comment (mccurdyc) [reply docs/network-setup.md:135]: @dianpopa sorry I don't understand. Are you suggesting to removing the space between the "On The Guest" section header and the first numbered bullet (i.e., "Define an unused IP...")?

I'm confused because I thought I saw in CI that it expected whitespace surrounding bullets.

Comment (mccurdyc) [reply docs/network-setup.md:133]: @dianpopa I gave it a shot! https://github.com/firecracker-microvm/firecracker/pull/2688/commits/a34e108c70fdd0bc784430b8ed619a083cc297e9

If you re-run CI, we can see and get it working :)

Comment (acatangiu) [reply docs/network-setup.md:133]: @mccurdyc you can easily test the markdown linter on your local or remote setup, no need to rely on running our full CI upstream:
```bash
tools/devtool test -- integration_tests/style/test_markdown.py
```

Or you could run all "style" tests:
```bash
tools/devtool test -- integration_tests/style/
```

Comment (mccurdyc) [reply docs/network-setup.md]: Wow I feel silly now. I forgot that I hard wrapped that line because there was a line-length markdown error. Now your original comment makes sense and I've pushed up a commit that should address it. I've gone ahead and squashed my commits. Thank for being patient with me @acatangiu @dianpopa

**Review by dianpopa — APPROVED** [review]

**Review by serban300 — APPROVED** [review]
